src/target/riscv/riscv.c

```diff
@@ -477,203 +557,227 @@ static int maybe_add_trigger_t1(struct target *target,
+		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			goto MATCH_EQUAL; /* Fallback to EQUAL MATCH */
+		return ret;
+	}
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_MATCH, CSR_MCONTROL6_MATCH_LT);
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_CHAIN, CSR_MCONTROL6_CHAIN_DISABLED);
+	tdata2 = trigger->address + trigger->length;
+	ret = set_trigger(target, idx + 1, tdata1, tdata2);
+	if (ret != ERROR_OK) {
+		set_trigger(target, idx, 0, 0); /* Undo the setting of the previous trigger*/
+		if (ret == ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
+			goto MATCH_EQUAL; /* Fallback to EQUAL MATCH */
+		return ret;
+	}
+	r->trigger_unique_id[idx] = trigger->unique_id;
+	r->trigger_unique_id[idx + 1] = trigger->unique_id;
+	return ERROR_OK;
 
+MATCH_EQUAL:
+	ret = find_trigger(target, CSR_TDATA1_TYPE_MCONTROL6, false, &idx);
+	if (ret != ERROR_OK)
+		return ret;
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_MATCH, CSR_MCONTROL6_MATCH_EQUAL);
+	if (trigger->length == 1)
+		tdata1 = set_field(tdata1, CSR_MCONTROL6_SIZE, CSR_MCONTROL6_SIZE_8BIT);
+	tdata1 = set_field(tdata1, CSR_MCONTROL6_CHAIN, CSR_MCONTROL6_CHAIN_DISABLED);
+	tdata2 = trigger->address;
+	ret = set_trigger(target, idx, tdata1, tdata2);
+	if (ret != ERROR_OK)
+		return ret;
+	r->trigger_unique_id[idx] = trigger->unique_id;
 	return ERROR_OK;
 }
 
 static int add_trigger(struct target *target, struct trigger *trigger)
 {
-	RISCV_INFO(r);
+	int ret;
+	riscv_reg_t tselect;
 
 	if (riscv_enumerate_triggers(target) != ERROR_OK)
 		return ERROR_FAIL;
 
-	riscv_reg_t tselect;
-	if (riscv_get_register(target, &tselect, GDB_REGNO_TSELECT) != ERROR_OK)
-		return ERROR_FAIL;
-
-	unsigned int i;
-	for (i = 0; i < r->trigger_count; i++) {
-		if (r->trigger_unique_id[i] != -1)
-			continue;
-
-		riscv_set_register(target, GDB_REGNO_TSELECT, i);
-
-		uint64_t tdata1;
-		int result = riscv_get_register(target, &tdata1, GDB_REGNO_TDATA1);
-		if (result != ERROR_OK)
-			return result;
-		int type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
-
-		result = ERROR_OK;
-		switch (type) {
-			case 1:
-				result = maybe_add_trigger_t1(target, trigger, tdata1);
-				break;
-			case 2:
-				result = maybe_add_trigger_t2(target, trigger, tdata1);
-				break;
-			case 6:
-				result = maybe_add_trigger_t6(target, trigger, tdata1);
-				break;
-			default:
-				LOG_DEBUG("trigger %d has unknown type %d", i, type);
-				continue;
-		}
-
-		if (result != ERROR_OK)
-			continue;
+	int result = riscv_get_register(target, &tselect, GDB_REGNO_TSELECT);
+	if (result != ERROR_OK)
+		return result;
 
-		LOG_DEBUG("[%d] Using trigger %d (type %d) for bp %d", target->coreid,
-				i, type, trigger->unique_id);
-		r->trigger_unique_id[i] = trigger->unique_id;
-		break;
-	}
+	do {
+		ret = maybe_add_trigger_t1(target, trigger);
+		if (ret == ERROR_OK)
+			break;
+		ret = maybe_add_trigger_t2(target, trigger);
+		if (ret == ERROR_OK)
+			break;
+		ret = maybe_add_trigger_t6(target, trigger);
+		if (ret == ERROR_OK)
+			break;
```

Comment:
This leads to some pretty poor behavior if only type 6 triggers are implemented. The code will end up writing tselect, and reading tdata2 twice for every trigger (once for t1 and once for t2). And that happens every time a trigger is set, which means every time you resume execution from gdb.

The best solution is probably to store the supported types in riscv_enumerate_triggers(), then there's no need to access the target to check the supported type at all.

Reply:
The new patch adds type records in riscv_enumerate_triggers, so finding triggers no longer requires writing tselect and reading tdata1